Peg-in to any address (#170)
rsksmart/RSKIPs: IPs/RSKIP170.md

# Peg-in to any address

|RSKIP          |170           |
| :------------ |:-------------|
|**Title**      |Peg-in to any address |
|**Created**    |01-SEP-20 |
|**Author**     |MI |
|**Purpose**    |USa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

This RSKIP proposes a way for users to be able to lock funds from BTC, indicating the RSK address in which they want their RBTCs transferred to and a BTC refund address in case the peg-in can not be completed for some reason.

## Motivation

Allowing users to lock funds to any address in RSK.

## Specification

The BTC transaction sent to the federation must contain at most one output with value 0 and OP_RETURN op code followed by the following data:
- Protocol version
- A valid RSK address (20 bytes)

> - A valid BTC refund address (optional)

Reviewer: Specify length of all the fields please

Reviewer 2: Please indicate any other restrictions (if required) for the transaction to be processed such as:
- Maximum length in bytes of transaction
- Order of OP_RETURN in outputs
- Maximum number of outputs

If no restriction is necessary, then explain why.

Reviewer 2: Also explain what is the size and value of protocol version

Author: Added a more detailed explanation, with the corresponding restrictions. Let me know if it's not clear, or if a restriction should be set to the number of outputs

## Rationale

> Discuss design decisions, community debates and possible attacks.

Reviewer 2: Explain why this RSKIP supersedes RSKIP41.

Reviewer 2: Explain what happens if the target address is a contract instead of an account.

Author: Added explanation. Let me know if you think the behaviour should be different when the destination address is a contract

# Peg-in to any address

|RSKIP          |170           |
| :------------ |:-------------|
|**Title**      |Peg-in to any address |
|**Created**    |01-SEP-20 |
|**Author**     |MI |
|**Purpose**    |USa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

This RSKIP proposes a way for users to be able to lock funds from BTC, indicating the RSK address in which they want their RBTCs transferred to and a BTC refund address in case the peg-in can not be completed for some reason.

## Motivation

Allowing users to lock funds to any address in RSK.

> ## Specification

Reviewer: I think you could give an idea on how this would be included in the existing flow:
1. Blockchain verifications (not modified)
2. parsing
2.1. rejection if parsing is invalid (following existing rules of refund)
2.2. flow if there is no op_return (not modified)
3. flow if op_return is present and valid

Reviewer 3: Abstract could start by: _This RSKIP allows a user to transfer BTC to RSK_, to make it more clear

The BTC transaction sent to the federation must contain at most one output with value 0 and OP_RETURN op code followed by the following data:

> - Protocol version (2 bytes). Currently **1**

Reviewer: ```suggestion
- Protocol version (2 bytes). Currently **01**
```

- A valid RSK address (20 bytes)
- [Optional] A valid BTC refund address in the following format:
    - Address type (1 byte)
        - 1 for P2PKH
        - 2 for P2SH
    - Hash needed to build refund address (20 bytes)
        - For P2PKH type address: public key hash 
        - For P2SH type address: script hash

### Example output
```
{
      "value": 0.00000000,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_RETURN 00010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "hex": "6a2b00010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "type": "nulldata"
      }
    }
```
Payload included: `00010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

- First 2 bytes correspond to the version number: `0001`
- Next 20 bytes indicate rsk destination address: `0e537aad84447a2c2a7590d5f2665ef5cf9b667a`
- Next 1 byte indicating btc refund address type, in this case P2PKH: `01`
- Finally 20 bytes indicating the public key hash that will be used to get the btc refund address: `4f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

### Notes

> - Only one output with OP_RETURN op code is allowed in the pegin transaction. Transactions with more than one OP_RETURN outputs will be rejected.

Reviewer 2: I think that rejecting in case there are more than one OP_RETURN is a bad idea. This is because there are use cases related with Mining products (i.e. decentralized pools) that we want to send bitcoins directly from coinbase transactions to a smart contract. At the same time other blockchains require OP_RETURNs to merge-mine with Bitcoin, so it won't be possible to do both at the same time, losing revenue.

Therefore I suggest that before protocol version, we add a 4-byte magic number (.i.e RSKT) for RSK Transfer. Then only forbid more than one OP_RETURN  with the RSKT magic world.

> - Only one output with OP_RETURN op code is allowed in the pegin transaction. Transactions with more than one OP_RETURN outputs will be rejected.

Author: Added specification that `RSKT` in ascii and hexa encoded (`52534b54`) should be added as prefix to the payload in OP_RETURN output

- No restriction is added to the maximum length in bytes of the transaction.
- There is no limit to the amount of outputs the transaction can have, as long as there is no more than one with OP_RETURN op code.
- The output with OP_RETURN does not have to be in any particular order among the other outputs.

## Rationale

[RSKIP41](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP41.md) includes optional fields with the purpose of executing a smart contract in RSK, in case the RSK destination address is a contract instead of an account. In this version of the protocol, the possibility of executing a smart contract from the pegin transaction is not contemplated. If the RSK destination address set in the payload is a contract wallet then the funds locked will be transferred to fund this contract, if the contract is not wallet then those funds will be lost.

> 

Reviewer 2: There is no definition of what a "contract wallet" is.

And also "funds will be lost" doesn't mean anything. Funds can be burned. Lost may also mean funds can end up in some other place, like a random contract getting the funds.
The logic must be clear. There are two types of accounts: 
- EOC
- Contracts

Contracts are identified because they have STORAGE.

Will contract code be executed ? Will the transfer happen without executing code (as it happens with REMAC payments?)

Author: Included specification that funds will be transferred to either EOA or contracts. In case of a contract, no code will be executed.

Reviewer 2: thanks for all the corrections. Looks good now.

Reviewer: @SergioDemianLerner looks like you commented that the rskip is ok now but the PR still has this as a **change request** from you.
Could you review it again please?
Thanks!

# Peg-in to any address

|RSKIP          |170           |
| :------------ |:-------------|
|**Title**      |Peg-in to any address |
|**Created**    |01-SEP-20 |
|**Author**     |MI |
|**Purpose**    |USa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

This RSKIP proposes a way for users to transfer BTC to RSK, indicating the RSK address in which they want their RBTCs transferred to and a BTC refund address in case the peg-in can not be completed for some reason.

## Motivation

> Allowing users to lock funds to any address in RSK, EOA or contract.

Reviewer: ```suggestion
Allowing users to lock peg-in BTC to any address in RSK, EOA or contract.
```

## Specification

This RSKIP proposes an extension to the current peg-in workflow. By including an output with an OP_RETURN op code and certain payload in the transaction sent to the federation to perform the peg-in, the user can determine the RSK address where the RBTCs will be transferred to and a BTC refund address. 

The BTC transaction sent to the federation must contain 
 one output with value 0 and OP_RETURN op code followed by the following data:
- `52534b54` as prefix indicating the output is for RSK. This values stands for `RSKT` (for RSK  transfer) in ascii and hex encoded. (4 bytes)

> - Protocol version in hexa (2 bytes). Currently **01**

Reviewer 2: Why in hexa and not a single binary byte?

Reviewer 2: If there is a strong reason, it should be in the rationale part

Author: To allow having multiple versions of the protocol, as we add new functionalities (contract execution, UTXOs consolidation, whatever may come next). We can reduce this field to just 1 byte, it still allows us to have 16 different protocol versions. More than enough I think.

        - 02 for P2SH
    - Hash needed to build refund address (20 bytes)
        - For P2PKH type address: public key hash 
        - For P2SH type address: script hash

### Example output
```
{
      "value": 0.00000000,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_RETURN 52534b5400010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "hex": "6a2b52534b54010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "type": "nulldata"
      }
    }
```
Payload included: `52534b54010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

- First 4 bytes correspond to the prefix that indicates the output has peg-in instructions for RSK (`RSKT` in hexa): `52534b54` 
- Next 1 byte that correspond to the version number in hexa: `01`
- Next 20 bytes indicate rsk destination address: `0e537aad84447a2c2a7590d5f2665ef5cf9b667a`
- Next 1 byte indicating btc refund address type, in this case P2PKH: `01`
- Finally 20 bytes indicating the public key hash that will be used to get the btc refund address: `4f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

### New peg-in flow

> When a BTC transaction is sent to the Bridge by a federate node, the following steps take place:

Reviewer: ```suggestion
When a BTC transaction is sent to the Bridge, the following steps take place:
```

1. Blockchain validations as performed in the existing flow (e.g. amount of confirmations, PMT valid)
2. Parse transaction to get RSK destination address and BTC refund address
   1. If no OP_RETURN data is present, follow the current flow. Get tx sender to use as BTC refund address and derive RSK destination address from BTC public key.
   2. If OP_RETURN data is present, get RSK destination address from the payload and the BTC refund address if present. In case there is no BTC refund address included in the payload, get the tx sender.

### Notes
- There is no limit to the amount of outputs with OP_RETURN the peg-in transaction can have. But there can be only one with `RSKT` prefix indicating peg-in instructions. Transactions with more than one OP_RETURN outputs containing `RSKT` prefix will be rejected, following the existing protocol for refunds.
- No restriction is added to the maximum length in bytes of the transaction.

> - There is no limit to the amount of outputs the transaction can have, as long as there is no more than one with OP_RETURN op code containing `RSKT` prefix in its payload.

Reviewer: This note seems duplicated with the first one

Author: Actually, this one refers to the total amount of outputs while the first one refers to the amount of outputs with OP_RETURN. I agree it is not clear, I changed it now

# Peg-in to any address

|RSKIP          |170           |
| :------------ |:-------------|
|**Title**      |Peg-in to any address |
|**Created**    |01-SEP-20 |
|**Author**     |MI |
|**Purpose**    |USa |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

This RSKIP proposes a way for users to be able to lock funds from BTC, indicating the RSK address in which they want their RBTCs transferred to and a BTC refund address in case the peg-in can not be completed for some reason.

## Motivation

Allowing users to lock funds to any address in RSK, EOA or contract.

## Specification

> This RSKIP proposes an extension to the current pegin workflow. By including an output with an OP_RETURN op code and certain payload in the transaction sent to the federation to perform the peg-in, the user can determine the RSK address where the RBTCs will be transferred to. 

Reviewer 3: You could add at the end: _the user can determine the RSK address where the RBTCs will be transferred to and a BTC refund address._

The BTC transaction sent to the federation must contain 
 one output with value 0 and OP_RETURN op code followed by the following data:

> - `RSKT` (for RSK  transfer) as prefix indicating the output is for RSK (4) bytes.

Reviewer: I think the idea is to store the result of `bytes4(hash(RSKT))`. If that's correct please specify the expected value en hexadecimal

- Protocol version (2 bytes). Currently **01**
- A valid RSK address (20 bytes)
- [Optional] A valid BTC refund address in the following format:
    - Address type (1 byte)
        - 1 for P2PKH
        - 2 for P2SH
    - Hash needed to build refund address (20 bytes)
        - For P2PKH type address: public key hash 
        - For P2SH type address: script hash

### Example output
```
{
      "value": 0.00000000,
      "n": 1,
      "scriptPubKey": {
        "asm": "OP_RETURN 52534b5400010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "hex": "6a2b52534b5400010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2",
        "type": "nulldata"
      }
    }
```
Payload included: `52534b5400010e537aad84447a2c2a7590d5f2665ef5cf9b667a014f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

> - First 4 bytes correspond to the prefix that indicates the output has peg-in instructions for RSK (`RSKT` in hexa): `52534b54` 

Reviewer: reading this it seems that what you are doing is getting the ASCII values of `RSKT`. If that's the idea specify it above please

- Next 2 bytes correspond to the version number: `0001`
- Next 20 bytes indicate rsk destination address: `0e537aad84447a2c2a7590d5f2665ef5cf9b667a`
- Next 1 byte indicating btc refund address type, in this case P2PKH: `01`
- Finally 20 bytes indicating the public key hash that will be used to get the btc refund address: `4f4c767a2d308eebb3f0f1247f9163c896e0b7d2`

### New peg-in flow

> This RSKIP proposes an extension of the current peg-in flow. When a BTC transaction is sent to the Bridge by a federate node, the following steps take place:

Reviewer: ```suggestion
When a BTC transaction is sent to the Bridge by a federate node, the following steps take place:
```
It seems redundant to talk about what the RSKIP proposes

> 1. Blockchain validations (amount of confirmations, PMT valid)

Reviewer: ```suggestion
1. Blockchain validations as performed in the existing flow (e.g. amount of confirmations, PMT valid)
```

2. Parse transaction to get RSK destination address and BTC refund address
   1. If no OP_RETURN data is present, follow the current flow.
   - Get tx sender to use as BTC refund address and derive RSK destination address from BTC public key. 
   - If the locked amount exceeds max locking cap or the sender is a multisig address, refund.
   - If the sender is a bech32 type address, reject the transaction.
   2. If OP_RETURN data is present, get RSK destination address from the payload and the BTC refund address if present. In case there is no BTC refund address included in the payload, get the tx sender.

>    - If no BTC refund address is included in the payload and the sender is a bech32 type address, then the funds can't be refunded. The lock will be attempted anyway.

Reviewer: You should mention that this flow will also check the locking cap
